desktopexporter/internal/app/utils/duration.ts

```diff
@@ -1,76 +1,106 @@
-import Timestamp from "timestamp-nano";
 import { SpanData } from "../types/api-types";
-
-export type TraceTiming = {
-  traceStartTimeNS: number;
-  traceDurationNS: number;
+import { PreciseTimestamp } from "../types/precise-timestamp";
+
+export type Duration = {
+  startTime: PreciseTimestamp
+  endTime: PreciseTimestamp
+  milliseconds: number
+  nanoseconds: number
+  label: string
 };
 
-export function calculateTraceTiming(spans: SpanData[]): TraceTiming {
+export function getTraceDuration(spans: SpanData[]): Duration {
   if (!spans.length) {
     return {
-      traceStartTimeNS: 0,
-      traceDurationNS: 0,
+      startTime: new PreciseTimestamp(0, 0),
+      endTime: new PreciseTimestamp(0, 0),
+      milliseconds: 0,
+      nanoseconds: 0,
+      label: ""
     };
   }
 
-  let earliestStartTime = getNsFromString(spans[0].startTime);
-  let latestEndTime = getNsFromString(spans[0].endTime);
+  let earliestStartTime = spans[0].startTime;
+  let latestEndTime = spans[0].endTime;
 
   spans.forEach((span) => {
-    let spanStart = getNsFromString(span.startTime);
-    if (spanStart < earliestStartTime) {
+    let spanStart = span.startTime;
+    if (spanStart.isBefore(earliestStartTime)) {
       earliestStartTime = spanStart;
     }
 
-    let spanEnd = getNsFromString(span.endTime);
-    if (spanEnd > latestEndTime) {
+    let spanEnd = span.endTime;
+    if (spanEnd.isAfter(latestEndTime)) {
       latestEndTime = spanEnd;
     }
   });
 
-  return {
-    traceStartTimeNS: earliestStartTime,
-    traceDurationNS: latestEndTime - earliestStartTime,
-  };
+  return getDuration(earliestStartTime, latestEndTime);
 }
 
-export function getNsFromString(timestampString: string): number {
-  let milliseconds = Date.parse(timestampString.split(".")[0]);
-  let nanoseconds =
-    milliseconds * 1e6 + Timestamp.fromString(timestampString).getNano();
-  return nanoseconds;
-}
-
-export function getDurationNs(startTimestamp: string, endTimestamp: string) {
-  let startTimeNs = getNsFromString(startTimestamp);
-  let endTimeNs = getNsFromString(endTimestamp);
+export function getDuration(startTime: PreciseTimestamp, endTime: PreciseTimestamp): Duration {
+  if (startTime === null || endTime === null) {
+    return {
+      startTime: startTime,
+      endTime: endTime,
+      milliseconds: 0,
+      nanoseconds: 0,
+      label: ""
+    };
+  }
 
-  return endTimeNs - startTimeNs;
-}
+  // Calculate duration in milliseconds and nanoseconds separately
```

Comment:
This logic likely belongs in the `PreciseTimestap` class, maybe as a class method. It's tied to the internal representation of the class.